From 594cdf6cc67feae584765c60a5261d0d8b6b9b45 Mon Sep 17 00:00:00 2001
From: Henry Schreiner <henryschreineriii@gmail.com>
Date: Wed, 29 Jan 2020 11:30:16 -0500
Subject: [PATCH] Allow shared CUDA to be linked

---
 CMakeLists.txt | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3b11ffd48..f7adf4cbc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -191,6 +191,7 @@ endif()
 
 if(CMAKE_VERSION VERSION_LESS 3.8)
     set(NEW_CUDA OFF CACHE INTERNAL "")
+    option(GOOFIT_CUDA_SHARED "Enable shared linking to CUDA")
 else()
     if(NOT DEFINED NEW_CUDA)
         set(NEW_CUDA ON CACHE INTERNAL "")
@@ -312,7 +313,34 @@ endif()
 # It is also used to setup up the NEW CUDA support
 macro(GOOFIT_OPTIONAL_CUDA)
     if(NEW_CUDA AND GOOFIT_DEVICE STREQUAL CUDA)
+
+        # Workaround until CMake 3.17 is released,
+        # see https://gitlab.kitware.com/cmake/cmake/issues/17559
+
+        if(GOOFIT_CUDA_SHARED)
+            set(CMAKE_CUDA_FLAGS "" CACHE STRING "")
+            if (CMAKE_CUDA_FLAGS)
+                list(REMOVE_ITEM CMAKE_CUDA_FLAGS "-cudart static")
+            endif()
+            string(APPEND CMAKE_CUDA_FLAGS "-cudart shared")
+        endif()
+
         enable_language(CUDA)
+
+        if(GOOFIT_CUDA_SHARED)
+            if(CMAKE_CUDA_HOST_IMPLICIT_LINK_LIBRARIES)
+                list(REMOVE_ITEM CMAKE_CUDA_HOST_IMPLICIT_LINK_LIBRARIES "cudart_static")
+                list(REMOVE_ITEM CMAKE_CUDA_HOST_IMPLICIT_LINK_LIBRARIES "cudadevrt")
+                list(APPEND CMAKE_CUDA_HOST_IMPLICIT_LINK_LIBRARIES "cudart")
+
+            endif()
+            if(CMAKE_CUDA_IMPLICIT_LINK_LIBRARIES)
+                list(REMOVE_ITEM CMAKE_CUDA_IMPLICIT_LINK_LIBRARIES "cudart_static")
+                list(REMOVE_ITEM CMAKE_CUDA_IMPLICIT_LINK_LIBRARIES "cudadevrt")
+                list(APPEND CMAKE_CUDA_IMPLICIT_LINK_LIBRARIES "cudart")
+            endif()
+        endif()
+
         set(CMAKE_CUDA_STANDARD ${GOOFIT_CXX_CUDA_STANDARD})
         set(CMAKE_CUDA_STANDARD_REQUIRED ON)
         set(CMAKE_CUDA_EXTENSIONS OFF)
